File: Java/Recover_BST.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    private TreeNode wrongNode1;
    private TreeNode wrongNode2;
    private TreeNode visited;
    public void recoverTree(TreeNode root) {
        wrongNode1 = null;
        wrongNode2 = null;
        visited = null;
        discoverWrong(root);
        if (wrongNode2 == null) wrongNode2 = visited;
        if (wrongNode1 != null && wrongNode2 != null) {
            int tmp = wrongNode2.val;
            wrongNode2.val = wrongNode1.val;
            wrongNode1.val = tmp;
        }
    }
    private void discoverWrong(TreeNode root) {
        if (root == null) return;
        discoverWrong(root.left);
        if (wrongNode1 == null && visited != null && root.val < visited.val) {
            wrongNode1 = visited;
        }
        if (wrongNode1 != null && visited.val < wrongNode1.val && root.val >= wrongNode1.val) {
            wrongNode2 = visited;
        }
        visited = root;
        discoverWrong(root.right);
    }
}

File: C++/Recover_BST.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {

private:
    TreeNode *wrongNode1;
    TreeNode *wrongNode2;
    TreeNode *visited;
    void discoverWrong(TreeNode *root) {
        if (root == NULL) return;
        discoverWrong(root->left);
        if (wrongNode1 == NULL && visited != NULL && visited->val > root->val) {
            wrongNode1 = visited;
        }
        if (wrongNode1 != NULL && visited->val < wrongNode1->val && root->val >= wrongNode1->val) {
            wrongNode2 = visited;
        }
        visited = root;
        discoverWrong(root->right);
    }

public:
    void recoverTree(TreeNode *root) {
        wrongNode1 = NULL;
        wrongNode2 = NULL;
        visited = NULL;
        discoverWrong(root);
        if (wrongNode2 == NULL) wrongNode2 = visited;
        if (wrongNode1 != NULL && wrongNode2 != NULL) {
            int tmp = wrongNode2->val;
            wrongNode2->val = wrongNode1->val;
            wrongNode1->val = tmp;
        }
    }
};
